### PR TITLE
feat(Data/Int): port data.int.basic from the lean3 prelude

### PR DESCRIPTION
**Pull request by denayd**
I needed some facts about the integers and Lean4 core just defines them and leaves it at that. This file proves the ring properties, provides some lemmas about the Nat to Int coercion, and defines a couple extra functions like sign and gcd.
It has to slightly modify Algebra.Group.Defs because that also defines the ℤ notation

Next up: my port of Data.Int.Order which depends on #18

### CODE REVIEW

File: Mathlib/Data/Int/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+      simp
+      rw [Nat.mul_sub_left_distrib]
+
+
+
+lemma neg_of_nat_add (m n : ℕ) : neg_of_nat m + neg_of_nat n = neg_of_nat (m + n) := by
+  cases m
+  cases n
+  simp
+  simp [Nat.zero_add]
```
Comment (kbuzzard) [inline Mathlib/Data/Int/Basic.lean:460]: Can you present this in such a way that it's clearer which goal is being worked on here? e.g. using indentation? You could even just put m and n after the colon and split on them using the equation compiler

File: Mathlib/Data/Int/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+| (ofNat m), -[1+ n],   (ofNat k) => by simp [neg_of_nat_eq_sub_nat_nat_zero] rw [Int.add_comm, ← sub_nat_nat_add] rfl
+| (ofNat m), -[1+ n],   -[1+ k]   => by simp rw [← Nat.left_distrib, succ_add] rfl
+| -[1+ m],   (ofNat n), (ofNat k) => by simp [Nat.mul_comm] rw [← Nat.right_distrib, Nat.mul_comm]
+| -[1+ m],   (ofNat n), -[1+ k]   => by simp [neg_of_nat_eq_sub_nat_nat_zero] rw [Int.add_comm, ← sub_nat_nat_add] rfl
+| -[1+ m],   -[1+ n],   (ofNat k) => by simp [neg_of_nat_eq_sub_nat_nat_zero] rw [← sub_nat_nat_add] rfl
+| -[1+ m],   -[1+ n],   -[1+ k]   => by simp rw [← Nat.left_distrib, succ_add] rfl
+
+protected lemma distrib_right (a b c : ℤ) : (a + b) * c = a * c + b * c := by
+  rw [Int.mul_comm, Int.distrib_left]
+  simp [Int.mul_comm]
```
Comment (kbuzzard) [inline Mathlib/Data/Int/Basic.lean:506]: Can you do it with one `simp`?

**Review by kbuzzard — COMMENTED** [review]
I am not sure that the naming convention is the Lean 4 one, but I am not very good at Lean 4 naming conventions yet.

File: Mathlib/Data/Int/Basic.lean (hunk 2 above)
Comment (denayd) [reply Mathlib/Data/Int/Basic.lean]: yes! good spot.

**Comment by denayd** [comment]
> I am not sure that the naming convention is the Lean 4 one, but I am not very good at Lean 4 naming conventions yet.

Honestly I'm clueless as to what the Lean 4 naming convention is, or if there even is one. I just made snake_case aliases for things because that's what Nat.Basic does and left everything else the same as lean3. Do you know if there's some kind of document or zulip thread I can read that will clarify it a little?

**Comment by digama0** [comment]
The main zulip thread on the topic is https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/rfc.3A.20theorem.20names . I was pretty happy with the proposal [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/rfc.3A.20theorem.20names/near/240134052) and it seemed to be gaining traction until Leo vetoed it in the final hour, and the discussion hasn't progressed since then.

Lean 4 core has a mashup of naming conventions, because the convention changed over time and the old stuff (especially proofs) was not updated. Those snake_case aliases are my attempt to bring some order to the madness with the naming convention called "proposal 3" in the thread: PascalCase namespaces, modules, and types, javaCase functions, and snake_case theorems. The proposal variation of uncertain outcome is whether to javaCase propositions as well.

**Comment by avigad** [comment]
I'll add that I also like Mario's proposal.

**Comment by avigad** [comment]
@denayd Git reports a conflict when I try to rebase and merge. Could you rebase and fix it, and then I'll merge?

**Comment by avigad** [comment]
@denayd Oh, I could squash-and-merge -- never mind. Sorry for the noise.

**Pull request merged**